### PR TITLE
[Cart][Inventory] Verify quantity of units on adding to cart 

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | #5899
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 1]
```diff
@@ -209,4 +210,30 @@ public function iShouldSeeTheProductWithPrice($productName, $productPrice)
...
+     * @Then /^I should not be notified that (this product) does not have sufficient stock$/
+     */
+    public function iShouldNotBeNotifiedThatThisProductDoesNotHaveSufficientStock(ProductInterface $product)
+    {
+        Assert::false(
+            $this->showPage->hasProductOutOfStockValidationMessage($product),
+            sprintf('I should see validation message for %s product', $product->getName())
+        );
+
+        sleep(5);
```
Comment (tuka217) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: `waitFor` not working.

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 2]
```diff
@@ -140,7 +165,9 @@ protected function getDefinedElements()
             'attributes' => '#sylius-product-attributes',
             'name' => '#sylius-product-name',
             'out-of-stock' => '#sylius-product-out-of-stock',
-            'product_price' => '#product-price'
+            'product_price' => '#product-price',
+            'validation-errors' => '.sylius-validation-error',
+            'selecting-variants' => "#sylius-product-selecting-variant"
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: Alphabetical order [*]

File: src/Sylius/Bundle/ShopBundle/Resources/private/js/verification-of-quantity-during-adding-to-cart.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+                    $(document).find('#sylius-product-adding-to-cart > button').on('click', function() {
+                        return $(this).closest('form').addClass('loading');
+                    });
+
+                    $('#sylius-product-adding-to-cart').on('submit', function(event) {
+                        refresh(this, event);
+                    });
+                }
+                else {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/private/js/verification-of-quantity-during-adding-to-cart.js]: cs

File: src/Sylius/Behat/Context/Ui/CheckoutContext.php [hunk 4]
```diff
@@ -986,6 +986,17 @@ public function myOrderSCurrencyShouldBe($currencyCode)
     }
 
     /**
+     * @Given /^I should not be notified that (this product) does not have sufficient stock$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/CheckoutContext.php]: Then

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 5]
```diff
@@ -0,0 +1,23 @@
+@cart_inventory
+Feature: Prevent buying more products, than available in a stock
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Unnecessary comma.

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    Scenario: Not being able to add some product's items to the cart if the quantity is greater than amount of items in stock
+        When I add 6 products "T-shirt Mononoke" to the cart
+        Then I should still be on product "T-shirt Mononoke" page
+        And I should be notified that this product does not have sufficient stock
+
+    @ui @javascript
+    Scenario: Being able to add some product's items to the cart if the quantity is not greater than amount of items in stock
+        When I add 4 products "T-shirt Mononoke" to the cart
+        Then I should not be notified that this product does not have sufficient stock
+        And I should see my cart summary
```
Comment (Zales0123) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Why not use existing ``I should be on my cart summary page``? ``I see`` couple scenario to UI :/

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 7]
```diff
@@ -0,0 +1,23 @@
+@cart_inventory
+Feature: Prevent buying more products, than available in a stock
+    In order to buy only available quantity of product's items
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: In order to buy only available items?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    As a Visitor
+    I want to be prevented from adding product's items with quantity greater that amount of items available in the inventory
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: I want to be prevented from adding items over the available amount?

Comment (lchrusciel) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: `that` -> `then`

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt Mononoke" priced at "$12.54"
+        And "T-shirt Mononoke" product is tracked by the inventory
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Is this step even needed?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        And there are 5 items of product "T-Shirt Mononoke" available in the inventory
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: This step should automatically mark it as tracked by the inventory.

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php (hunk 1 above)
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: But why did you place it after assertion?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+    @ui @javascript
+    Scenario: Not being able to add some product's items to the cart if the quantity is greater than amount of items in stock
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Preventing from adding more items to the cart than it's available?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        When I add 6 products "T-shirt Mononoke" to the cart
+        Then I should still be on product "T-shirt Mononoke" page
+        And I should be notified that this product does not have sufficient stock
+
+    @ui @javascript
+    Scenario: Being able to add some product's items to the cart if the quantity is not greater than amount of items in stock
```
Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Allowing to add items to the cart if they are available?

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 13]
```diff
@@ -209,4 +210,30 @@ public function iShouldSeeTheProductWithPrice($productName, $productPrice)
             sprintf("The product %s with price %s should appear on page, but it does not.", $productName, $productPrice)
         );
     }
+
+    /**
+     * @Then /^I should be notified that (this product) does not have sufficient stock$/
+     */
+    public function iShouldBeNotifiedThatThisProductDoesNotHaveSufficientStock(ProductInterface $product)
+    {
+       $this->showPage->waitForValidationErrors(3);
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: This needs some refactor for sure 😱

Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: Sure it does :(

File: src/Sylius/Behat/Context/Ui/Shop/CartContext.php [hunk 14]
```diff
@@ -325,6 +325,7 @@ public function iAddProductsToTheCart($quantity, ProductInterface $product)
 
     /**
      * @Then I should be on my cart summary page
+     * @Then I should see my cart summary
```
Comment (Arminek) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: ``I should be ...`` WDYT?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature (hunk 10 above)
Comment (tuka217) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Why?

Comment (pamil) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: Even if it's possible to have 5 products in the inventory but not have them tracked, it still is rather unusual and obviously not the default choice. I'm just for merging these two steps into one.

Comment (tuka217) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: You have right

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 15]
```diff
@@ -140,7 +165,9 @@ protected function getDefinedElements()
             'attributes' => '#sylius-product-attributes',
             'name' => '#sylius-product-name',
             'out-of-stock' => '#sylius-product-out-of-stock',
-            'product_price' => '#product-price'
+            'product_price' => '#product-price',
+            'selecting-variants' => "#sylius-product-selecting-variant",
+            'validation-errors' => '.sylius-validation-error'
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: missing comma at the end

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    I want to be prevented from adding items over the available amount
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt Mononoke" priced at "$12.54"
+        And "T-shirt Mononoke" product is tracked by the inventory
+        And there are 5 items of product "T-Shirt Mononoke" available in the inventory
+
+    @ui @javascript
+    Scenario: Preventing from adding more items to the cart than it's available in stock
```
Comment (GSadee) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: ...in **a** stock...

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+        When I add 6 products "T-shirt Mononoke" to the cart
+        Then I should still be on product "T-shirt Mononoke" page
+        And I should be notified that this product does not have sufficient stock
+
+    @ui @javascript
+    Scenario: Allowing to add items to the cart if they are in stock
```
Comment (GSadee) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: ...in **a** stock...

Comment (pjedrzejewski) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: I believe ``in stock`` is correct?

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature (hunk 16 above)
Comment (pjedrzejewski) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: See my comment below. :)

**Comment by pjedrzejewski** [comment]
Thank you Ania, nice work! :)

File: features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature (hunk 17 above)
Comment (steffenbrem) [inline features/inventory/cart_inventory/prevent_buying_more_products_than_available_in_a_stock.feature]: `in stock` is correct indeed.

**Pull request merged**